src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala

```diff
@@ -233,26 +233,28 @@ private class TaskLauncherActor(
   }
 
   private[this] def receiveInstanceUpdate: Receive = {
-    case update: InstanceDeleted =>
-      logger.info(s"receiveInstanceUpdate: ${update.id} was deleted (${update.condition})")
-      // A) If the app has constraints, we need to reconsider offers that
-      // we already rejected. E.g. when a host:unique constraint prevented
-      // us to launch tasks on a particular node before, we need to reconsider offers
-      // of that node after a task on that node has died.
-      //
-      // B) If a reservation timed out, already rejected offers might become eligible for creating new reservations.
-      if (runSpec.constraints.nonEmpty || (runSpec.isResident && shouldLaunchInstances)) {
-        maybeOfferReviver.foreach(_.reviveOffers())
+    case update: InstanceUpdated =>
+      if (update.condition.isTerminal & update.instance.isScheduled) {
+        logger.info(s"receiveInstanceUpdate: ${update.id} is terminal (${update.condition}) and scheduled.")
+        // A) If the app has constraints, we need to reconsider offers that
+        // we already rejected. E.g. when a host:unique constraint prevented
+        // us to launch tasks on a particular node before, we need to reconsider offers
+        // of that node after a task on that node has died.
+        //
+        // B) If a reservation timed out, already rejected offers might become eligible for creating new reservations.
+        if (runSpec.constraints.nonEmpty || (runSpec.isResident && shouldLaunchInstances)) {
+          maybeOfferReviver.foreach(_.reviveOffers())
```

Comment:
it seems like we should not revive if the update goal is not running

Reply:
When do we revive for newly created instances?

Reply:
Relevant to a discussion Alena and I had: if we revive for newly created instances, then when a TaskReplaceActor marks some ephemeral instance as decommissioned, then watches for the instance to become terminal, the TaskReplaceActor should create a new instance with goal running, and this should ultimately trigger a revive.

Reply:
I'm not sure I understand. How is your example related to this logic? As far as I understand the comment we revive to see old rejected offers again because they might not be rejected this time. How does this related to newly created instances if their run spec did not change?

Reply:
What do you mean by ` run spec did not change` - how did that come in?

Reply:
You need only to revive offer if the run spec config ref is different, it hold constraints or it's a resident. So to answer Tim's question, if an instance is terminal and scheduled we might have to consider old offer, eg if the task was pinned to the host. Thus we have to revive.